artifact/v3/header/token.hpp

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_ARTIFACT_V3_HEADER_TOKEN_HPP
+#define MENDER_ARTIFACT_V3_HEADER_TOKEN_HPP
+
+
+#include <unordered_map>
+#include <string>
+#include <regex>
+
+#include <common/log.hpp>
+#include <common/expected.hpp>
+
+const std::regex artifact_script_regexp {
+	"scripts/Artifact(Install|Reboot|Rollback|RollbackReboot|Failure)_(Enter|Leave)_[0-9][0-9](_\\S+)?",
+	std::regex_constants::ECMAScript};
+
+namespace mender {
+namespace artifact {
+namespace v3 {
+namespace header {
+namespace token {
+
+using namespace std;
+
+namespace tar = mender::tar;
+
+namespace log = mender::common::log;
+namespace error = mender::common::error;
+namespace expected = mender::common::expected;
+
+enum class Type {
+	Uninitialized = 0,
+	EOFToken,
+	Unrecognized,
+	HeaderInfo,
+	ArtifactScripts,
+	ArtifactHeaderTypeInfo,
+	ArtifactHeaderMetaData,
+};
+
+unordered_map<const Type, string> type_map {
+	{Type::Uninitialized, "Uninitialized"},
+	{Type::EOFToken, "EOF"},
+	{Type::Unrecognized, "Unrecognized"},
+	{Type::HeaderInfo, "header-info"},
+	{Type::ArtifactScripts, ""}, // Will be set during parsing
+	{Type::ArtifactHeaderTypeInfo, "header-type-info"},
+	{Type::ArtifactHeaderMetaData, "header-meta-data"},
+};
+
+class Token {
+public:
+	Type type;
+	shared_ptr<tar::Entry> value;
+
+public:
+	Token() :
+		type {Type::Uninitialized} {
+	}
+	Token(Type t) :
+		type {t} {
+	}
+	Token(const string &type_name, tar::Entry &entry) :
+		type {StringToType(type_name)},
+		value {make_shared<tar::Entry>(entry)} {
+	}
+
+	const string TypeToString() const {
+		return type_map.at(type);
+	}
+
+private:
+	Type StringToType(const string &type_name) {
+		if (type_name == "header-info") {
+			return Type::HeaderInfo;
+		}
+		if (std::regex_match(type_name, artifact_script_regexp)) {
+			const string script_name = type_name.substr(7 + 1); // Strip the scripts + / prefix
+			type_map[Type::ArtifactScripts] = script_name;
```

Comment:
Modifying a global `type_map` from a function in `Token` looks quite fishy.

Reply:
I'd have to agree. `type_map` is a singleton, but `Token` isn't, so this is a dangerous relationship.